AVM FRITZ!Box 7530: Add support for Toshiba NAND (8 bit ECC)

andyboeh opened the pull request:
Add support for FRITZ!7530 with Toshiba NAND (8 bit ECC)

Some revisions of the FRITZ!7530 use a Toshiba NAND with 8 bit ECC in contrast to the Macronix NAND with 4 bit ECC. Since the ECC information is hardcoded in qcom-ipq4019.dtsi, this restrictions is removed from the device DTS.

An additional kernel patch is necessary to identify the NAND chip. The autodetection misdetects 64 byte OOB where the NAND uses 128 byte. A proper fixup was already merged into the respective U-Boot code.

hauke commented on target/linux/ipq40xx/patches-5.4/0020-mtd-nand-Add-support-for-Toshiba-TC58NVG0S3HTA00.patch:
```diff
@@ -0,0 +1,12 @@
+ 		  2 },
++	{"TC58NVG0S3HTA00 1G 3.3V 8-bit",
++		{ .id = {0x98, 0xf1, 0x80, 0x15} },
++		  SZ_2K, SZ_128, SZ_128K, 0, 4, 128, NAND_ECC_INFO(8, SZ_512), },
```
Please move this to target/linux/generic/ for kernel 5.10 and 5.15.
 
It would be nice if you could send this upstream too.

andyboeh replied:
Ah damn, I forgot to sync my fork! I was unsure whether it should be in platform or generic, but I'll move it to generic. I wanted to send it upstream anyways.

hauke replied:
Could you already send this upstream and then add into the pending folder please. Please also add the git patch header you used upstream.

hauke commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530-toshiba.dts:
```diff
@@ -0,0 +1,92 @@
+
+	nand@0 {
+		nand-ecc-strength = <8>;
+		nand-ecc-step-size = <512>;
```
Is this the only difference between the two dts files?
Is it possible to use these settings on the older devices too?

andyboeh replied:
Yes. I took the VR9-based Fritz 3370 as template for the changes.
I don't think it will work on the older devices, but I don't have any to test. I think that the other NAND chips require 4 Bit ECC (in U-Boot, 4 Bit ECC is (was) hardcoded in the NAND controller driver and thus failed for this chip).
The (ugly) hack for U-Boot is here: https://github.com/chunkeey/FritzBox-4040-UBOOT/pull/8

hauke replied:
ok, thanks for the clarification

andyboeh committed:
kernel: add support for Toshiba TC58NVG0S3HTA00 NAND flash

The Toshiba TC58NVG0S3HTA00 is detected with 64 byte OOB while the flash
has 128 byte OOB. This adds a static NAND ID entry to correct this.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

andyboeh commented:
I added a few more files to the commit, I forgot to adapt the configuration files to the changed device IDs.

hauke commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -276,7 +276,28 @@ define Device/avm_fritzbox-7530
+	DEVICE_ALT0_VENDOR := AVM
+	DEVICE_ALT0_MODEL := FRITZ!Box 7520
+	SOC := qcom-ipq4019
+	DEVICE_PACKAGES := fritz-caldata fritz-tffs-nand
```
I think you should add a `SUPPORTED_DEVICES:=avm_fritzbox-7530` here to make it possible to upgrade old avm_fritzbox-7530 images to the macronix variant.

hauke commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7530.dtsi:
```diff
@@ -167,85 +167,6 @@
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			partition@0 {
```
Is the partition layout different between both devices? If it is the same you can also leave it in this file and only change the `nand-ecc-strength` and `nand-ecc-step-size` settings in the device specific dts files.

andyboeh commented:
@hauke 
I think I addressed all your comments with the last force-push. The MTD patch has been sent upstream and has been merged into mtd/next.
I added another commit that patches U-Boot (it's ugly, I know) until the upstream patch is merged at https://github.com/chunkeey/FritzBox-4040-UBOOT/pull/8

hauke commented:
This looks good to me
@chunkeey  Could you please look at the u-boot changes.

janh commented:
`BUILD_DEVICES` in the `uboot-fritz4040` package Makefile also needs to be updated with the new device names, so that it is automatically selected.

janh commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -276,7 +276,29 @@ define Device/avm_fritzbox-7530
 	SOC := qcom-ipq4019
 	DEVICE_PACKAGES := fritz-caldata fritz-tffs-nand
 endef
```
The old device section is still there, but I think it isn't needed anymore.

andyboeh replied:
You're right. Initially, I took the Fritz!3370 as a reference but then I did not include the original device section.

janh commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -276,7 +276,29 @@ define Device/avm_fritzbox-7530
+define Device/avm_fritzbox-7530-toshiba
+	$(call Device/FitImageLzma)
+	DEVICE_VENDOR := AVM
+	DEVICE_MODEL := FRITZ!Box 7530 (Toshiba NAND)
```
Using `DEVICE_VARIANT` instead for the NAND type feels more semantically correct.

janh commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -276,7 +276,29 @@ define Device/avm_fritzbox-7530
+	DEVICE_VENDOR := AVM
+	DEVICE_MODEL := FRITZ!Box 7530 (Toshiba NAND)
+	DEVICE_ALT0_VENDOR := AVM
+	DEVICE_ALT0_MODEL := FRITZ!Box 7520
```
Shouldn't the alternate device also include the NAND type?

andyboeh commented:
> `BUILD_DEVICES` in the `uboot-fritz4040` package Makefile also needs to be updated with the new device names, so that it is automatically selected.

Ah, that's why it wasn't built automatically! Can I add multiple calls to BUILD_DEVICES or is it an array of some kind? Do I need to add a second definition?

andyboeh commented:
@janh @hauke I think I addressed all issues.
@chunkeey Could you have a look at the U-Boot PR, please?

chunkeey commented:
Don't have a 7530, I really went up in price.

But ok, thank you for the uboot patch. I changed it to use a switch-case table. Please let me know if this works or if I broke something in the conversion. (Upstream linux has a big struct "nand_flash_ids" table too).

What I wonder though: You had a patch which added the ECC values into the kernel's nand_flash_ids table in drivers/mtd/nand/raw/nand_ids.c . What became of it? Because from what I think to remember is: If you have the values there, you don't need the nand-ecc properties in the dts (so if this entry there worked, you should be able to use the same dts as the 7530/7520).

andyboeh commented:
> What I wonder though: You had a patch which added the ECC values into the kernel's nand_flash_ids table in drivers/mtd/nand/raw/nand_ids.c . What became of it? Because from what I think to remember is: If you have the values there, you don't need the nand-ecc properties in the dts (so if this entry there worked, you should be able to use the same dts as the 7530/7520).

The patch in nand_ids.c has been merged upstream. However, the NAND ECC strength is overriden by the qcom-ipq4019.dtsi, therefore the separate DTS is necessary (see: https://github.com/torvalds/linux/blob/b2da7df52e16110c8d8dda0602db81c15711e7ff/arch/arm/boot/dts/qcom-ipq4019.dtsi#L503). If the value was not there by default, it would probably work.

It's going to take a week before I can test the updated U-Boot, thanks.

chunkeey commented:
> > What I wonder though: You had a patch which added the ECC values into the kernel's nand_flash_ids table in drivers/mtd/nand/raw/nand_ids.c . What became of it? Because from what I think to remember is: If you have the values there, you don't need the nand-ecc properties in the dts (so if this entry there worked, you should be able to use the same dts as the 7530/7520).
> 
> The patch in nand_ids.c has been merged upstream. However, the NAND ECC strength defined there is overriden by the qcom-ipq4019.dtsi, therefore the separate DTS is necessary (see: https://github.com/torvalds/linux/blob/b2da7df52e16110c8d8dda0602db81c15711e7ff/arch/arm/boot/dts/qcom-ipq4019.dtsi#L503). If the value was not there by default, it would probably work.

Thank you. It is possible to nuke those values by placing  [/delete-property/ node;](https://elinux.org/Device_Tree_Source_Undocumented#Deleting_nodes_and_properties) into the ```nand@0``` node.

    /delete-property/ nand-ecc-strength;
    /delete-property/ nand-ecc-step-size;

@blocktrron do you have a 7530 to test this with? Or would it be better to stick with two variants?
 
> It's going to take a week before I can test the updated U-Boot, thanks.
(There's also the issue with the bad MAC-Addresses in uboot there. hmm)

janh commented:
> It is possible to nuke those values by placing [/delete-property/ node;](https://elinux.org/Device_Tree_Source_Undocumented#Deleting_nodes_and_properties) into the `nand@0` node.
> 
> ```
> /delete-property/ nand-ecc-strength;
> /delete-property/ nand-ecc-step-size;
> ```

This works with my 7530 and 7520 (both are older devices with the Macronix MX30LF1G18AC NAND chip).



> (There's also the issue with the bad MAC-Addresses in uboot there. hmm)

Is there a separate issue somewhere to discuss the MAC address problem? (The MAC address is correct here as long as I use the matching u-boot variant. But the 7520 variant isn't built by the u-boot package.)

dhewg commented:
> Is there a separate issue somewhere to discuss the MAC address problem? (The MAC address is correct here as long as I use the matching u-boot variant. But the 7520 variant isn't built by the u-boot package.)

That variant was build with the initial patches of #4867, but all of that was dropped in favor of `DEVICE_ALT0_*`.

I'm not sure what the best approach is, maybe u-boot can be kept as a single variant and it auto-detect `CONFIG_AVM_EVA_CFG_OFFSET`?

andyboeh commented:
> > It is possible to nuke those values by placing [/delete-property/ node;](https://elinux.org/Device_Tree_Source_Undocumented#Deleting_nodes_and_properties) into the `nand@0` node.
> > ```
> > /delete-property/ nand-ecc-strength;
> > /delete-property/ nand-ecc-step-size;
> > ```
> 
> This works with my 7530 and 7520 (both are older devices with the Macronix MX30LF1G18AC NAND chip).

I'll test it on the weekend with my 7530. That would then mean that we don't need a separate target, right?

chunkeey commented:
> I'll test it on the weekend with my 7530. That would then mean that we don't need a separate target, right?

Yes, that's the idea. 

> I'm not sure what the best approach is, maybe u-boot can be kept as a single variant and it auto-detect CONFIG_AVM_EVA_CFG_OFFSET?

[From what I know](https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=commit;h=ed104fde0682751c21c0cc81b23ce5612922f0ce), the existing 7520/7530 (variants?) already have a different EVA_CFG_OFFSET location. In theory, a TFFS parser could be included in uboot. Though I think this is outside of the scope of this PR.

chunkeey commented:
FYI: updated uboot-fritz4040, so the uboot portion could be removed.

andyboeh committed:
IPQ4019: AVM FRITZ!Box 7530: Remove NAND ECC restrictions from DTS

Some revisions of the FRITZ!7530 use a Toshiba NAND with 8 bit ECC in
contrast to the Macronix NAND with 4 bit ECC. This removes the hardcoded
ECC strength and step size as set in qcom-ipq4019.dtsi, thus relying on the
kernel NAND detection routines to correclty set up the ECC parameters.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

andyboeh commented:
Finally, I had the time to test the new U-Boot and the proposed changes and I can confirm: Everything seems to work properly with the new U-Boot, the small kernel patch and the removal of the ECC restrictions from the DTS.

On my device, U-Boot and OpenWrt both read the MAC address correctly.

Thanks for all the input, this solution is much simpler and creates less confusion than creating a separate target.

chunkeey commented:
Merged. Thanks!
